// hikereactsdk/components/hikeContactGroupChooser/multiselect/styles.js
import React from 'react-native';

export default React.StyleSheet.create({
	list: {},

	row: {
		flexDirection: 'row',
		alignItems: 'center',
	},

	optionLabel: {
		flex: 1,
	},

	optionIndicator: {
    width: 48,
		height: 48,
		justifyContent: 'center',
		alignItems: 'flex-start'
	},

	optionIndicatorIcon: {
    width: 48,
		height: 48,
	},

	separator: {
		height: 1,
		marginTop: 4,
		marginBottom: 4,
		backgroundColor: '#efefef',
	},
	loader: {
		flex: 1,
		alignItems: 'center',
		justifyContent: 'center',
	},
	ctr: {
		flex: 1,
		justifyContent: 'flex-start',
	},
	sectionHeader:{
		justifyContent:'center',
		backgroundColor:'#ffffff'
	},
	listSeparator: {
		flex: 1,
		flexDirection: 'row',
		flexWrap: 'wrap',
	},

	emptyLeftSection: {
		width: 82,
		borderBottomWidth: 1,
		borderBottomColor: '#ffffff',
	},

	borderRow: {
		flex: 1,
		borderBottomWidth: 1,
		borderBottomColor: 'rgba(0,0,0,0.06)',
	},

	contactCategory: {
		alignSelf:'center',
		marginVertical: 18,
		fontSize:12,
		color:'rgba(0,0,0,0.3)'
	},

	contactItem: {
		flex:1,
		flexDirection: 'row',
		paddingLeft:16,
		alignItems: 'center',
		justifyContent:"center",
	},
	thumbnailContainer:{
		marginVertical:12,
	},
	contactIcon: {
		width: 48,
		height: 48,
		borderRadius: 24,
	},
  contactRightPane: {
    flex: 1,
		alignItems:"center",
		alignSelf:"center",
		justifyContent:"center",
    flexDirection: 'row',
    marginHorizontal: 16,
		borderBottomWidth: 1,
		borderColor: '#f4f4f4',
    alignSelf: 'stretch',
    alignItems: 'center',
  },
	contactDesc:{
		flex:1,
		height:48,
		justifyContent:"center"
	},
	contactName: {
		flex:1,
		flexDirection:'row',
		alignSelf:'flex-start',
//		fontFamily:"FaktSoftProMedium",
		fontSize: 16,
		color:'#3D475B',
		marginBottom:6
	},
	contactNumber:{
		flex:1,
		flexDirection:'row',
//		fontFamily:"FaktSoftProNormal",
		alignSelf:'flex-start',
		fontSize: 12,
		marginLeft: 5,
	},
	createCtaText:{
		color: '#1da8e8',
		fontSize:16,
//		fontFamily:"FaktSoftProMedium",
	}
});
