// hikereactsdk/components/contactPicker/ContactHeader.js
// Import libraries for making a component
import React, { Component } from 'react';
import {
  Text,
  View,
  Image,
  Platform,
  TextInput,
  TouchableOpacity
} from 'react-native';
import AppTheme from '../../appthemes/AppTheme';
import ColorUtils from '../../appthemes/ColorUtils';
import ThemeText from '../../appthemes/components/ThemeText';
import CustomTouch from '../CustomTouch';
import { Avatar } from '../Avatar';;
import IconStyles from '../../appthemes/IconStyles';
import ThemeIcon from '../../appthemes/components/ThemeIcon';
/**
* contact header accepts following props
* @property headerText
*      string to show the text on header
* @property onPress
*      callback on the press of back icon
* @property searchEnable
*      boolean to show the search icon and enable search functioning
* @property SearchFilterFunction
*      callback to search or do something on writing text in textInput
*/

class ContactHeader extends Component {
  constructor(props) {
    super(props);

    this.state = {
      show: false,
      searchText: '',
    }
    this.showHeaderText = true;
    this._searchPress = this._searchPress.bind(this);
    this._clear = this._clear.bind(this);
    this._onBack = this._onBack.bind(this);
  }

  componentWillMount() {
    if (this.props.rightEnable === undefined)
    this.props.rightEnable = false;
  }
  _searchPress() {
    this.setState({
      show: true,
    })
  }
  _onBack(){
        this._clear();
        this.setState({show:false}) ;
    }
  onChangeText(searchText) {
    this.setState({ searchText: searchText });
    this.props.SearchFilterFunction && this.props.SearchFilterFunction(searchText);

  }
  _clear() {
    this.setState({
      searchText: ''
    });
    this.props.SearchFilterFunction && this.props.SearchFilterFunction('');

  }
  render() {
    const {
      onPress,
      headerText,
      groupName,
      themeData,
      donePress,
      doneEnable,
      searchPress,
      searchEnable
    } = this.props;
    const {

      viewStyle,
      middle,
      Middle
    } = styles;
    var textColor = AppTheme.getColorPalette().transBgColor;
    var searchHintTextColor = ColorUtils.getColorWithOpacity(AppTheme.getColorPalette().transBgColor, 0.6);
    var imageColor=AppTheme.getColorPalette().mainColor;
    // var overrideThemeColor = this.props.themeType === 2 ? { color: themeData.secondaryColor } : {}
    return (
      <View style={viewStyle}>
        <View style={Platform.OS === 'ios' ? { flex: 1, flexDirection: "row", height: 30, alignItems: 'center' } : { flex: 1, flexDirection: "row", height: 30, alignItems: 'center' }}>

          <View >
            <ThemeIcon

              onPress={()=> {this.state.show ? this._onBack()  : this.props.onPress()}}
              iconProfile='IconProfile18'
              source={require('../../images/ic_med_back.png')} />
          </View>
          {
            !this.state.show && <View style={[middle, Platform.OS === 'ios' ? { alignItems: 'center', justifyContent: 'space-around', flex: 0.8 } : { alignItems: 'flex-start' }]} >
            <ThemeText
              text={headerText}
              color={textColor}
              fontProfile='FontProfile03'
              />
            <View >{

                groupName &&
                <ThemeText
                  text={groupName}
                  color={textColor}
                  fontProfile='FontProfile18'
                  />
              }
            </View>

          </View>
        }
        {this.state.show &&


          <View style={styles.searchCtr}>
            <TextInput style={[styles.search, { color: textColor }]}
              placeholder="Search by name or number"
              underlineColorAndroid="transparent"
              placeholderTextColor={searchHintTextColor}
              autoFocus={true}
              tintColor={ textColor}
              selectionColor={ textColor}
              onChangeText={(searchText) => { this.onChangeText(searchText); }}
              value={this.state.searchText} />

            {this.state.searchText.length ?

              <ThemeIcon
                onPress={this._clear}
                iconProfile='IconProfile18'
                source={require('../../images/ic_med_close.png')} />

              : null}
            </View>

          }
        </View>

        {
          doneEnable &&
          <TouchableOpacity onPress={donePress} >
            <ThemeText
              text={'Done'}
              color={textColor}
              fontProfile='FontProfile03'
              />
          </TouchableOpacity>
        }
        {
          !this.state.show && searchEnable &&
          <View >
            <ThemeIcon
              onPress={this._searchPress.bind(this)}
              iconProfile='IconProfile18'
              source={require('../../images/ic_med_search.png')} />
          </View>


        }
      </View>
    );
  }
};


const styles = {
  viewStyle: {
    flex: 1,
    flexDirection: "row",
    alignSelf: Platform.OS === 'ios' ? "center" : "flex-start",
    justifyContent: 'center',
    alignItems: 'center',
    height: 60,
    //backgroundColor:'#ff0000'
  },
  middle: {
    paddingLeft: 5
  },
  searchCtr: {

    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    height: 56
  },
  search: {
    marginHorizontal: 10,
    paddingHorizontal: 15,
    paddingVertical: 5,
    flex: 1,
    fontSize: 18,

  },


};





export default ContactHeader
