// hikereactsdk/components/contactPicker/ListRow.js
import React from 'react';
import { StyleSheet, Text, View, ToastAndroid, FlatList, Image, Platform, TouchableOpacity } from 'react-native';
import AppTheme from '../../appthemes/AppTheme';
import ThemeText from '../../appthemes/components/ThemeText';
import CustomTouch from '../CustomTouch';
import { Avatar } from '../Avatar';
import { Logger } from '../../index';
import ThemeIcon from '../../appthemes/components/ThemeIcon';

export default class ListRow extends React.Component {

  constructor(props) {
    super(props);
  }

  render() {
    Logger.debug("listrow is rendered");
    return (
      <CustomTouch onPress={this.props.onSelect}>
        <View style={styles.contactItem}>
          <Avatar name={this.props.row.name}
            thumbnail={this.props.row.thumbnail}
            size={48} />
          <View style={styles.contactDesc}>

            <View style={{ flex: 3 }}>
              <ThemeText style={styles.contactName}
                text={this.props.row.name}
                color={ AppTheme.getColorPalette().mainColor}
                fontProfile='FontProfile05'>
              </ThemeText>
              {this.props.row.members_count &&
              <View style={{flexDirection: 'row'}}>
                <ThemeIcon
                  style={{width:32,height:20,marginLeft:-8}}
                  color={AppTheme.getColorPalette().secondaryColor}
                  onPress={this.props.onSelect}
                  source={require('../../images/ic_med_group.png')}/>
                <ThemeText style={styles.contactName}
                  text={' '+this.props.row.members_count.split(' ')[0]}
                  color={AppTheme.getColorPalette().secondaryColor}
                  fontProfile='FontProfile19'>
                </ThemeText>
  						</View>
              }
            </View>

            {this.props.selected &&
              <View>
                <ThemeIcon
                  onPress={this.props.onSelect}
                  iconProfile='IconProfile03'
                  source={require('../../images/ic_med_tick.png')} />
              </View>
            }
          </View>
        </View>
      </CustomTouch>
    );
  }
}

const styles = StyleSheet.create({
  contactItem: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'flex-start',
    paddingLeft: 16,
    paddingTop: 12,
    paddingBottom: 12,
    paddingRight: 16,
    alignItems: 'center',
  },
  contactDesc: {
    flex: 4,
    flexDirection: 'row',
    alignItems: 'center',
    marginLeft: 16
  },
  contactName: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  contactIcon: {
    width: 48,
    height: 48,
    borderRadius: 24,
  }
})
