// hikereactsdk/components/contactPicker/ListHeader.js
import React from 'react';
import { StyleSheet, Text, View, ToastAndroid, FlatList, Image, Platform, TouchableOpacity } from 'react-native';
import AppTheme from '../../appthemes/AppTheme';
import ThemeText from '../../appthemes/components/ThemeText';
import CustomTouch from '../CustomTouch';
import { Avatar } from '../Avatar';
import ThemeIcon from '../../appthemes/components/ThemeIcon';

export default class ListHeader extends React.Component {

  constructor(props) {
    super(props);
  }

  render() {
    return (
      <CustomTouch onPress={this.props.onPress}>
        <View style={styles.container}>
          <View style={[styles.thumbnailContainer]}>
            <Image source={this.props.imag} style={styles.contactIcon} />
          </View>
          <View style={[styles.contactRightPane]}>
            <View style={[styles.contactDesc, { marginLeft: 0 }]}>
              <ThemeText
                style={styles.contactName}
                text={this.props.text}
                color={AppTheme.getColorPalette().accentColor}
                fontProfile='FontProfile04'
                />
            </View>
            {this.props.selected && <View>
              <ThemeIcon
                onPress={this.props.onPress}
                iconProfile='IconProfile03'
                source={require('../../images/ic_med_tick.png')} />
            </View>
          }
        </View>
      </View>
    </CustomTouch>
    )
  }
}
const styles = StyleSheet.create({
  container: {
    height: 72,
    flexDirection: 'row',
    paddingLeft: 16,
    alignItems: 'center',
    justifyContent: 'flex-start'
  },
  thumbnailContainer: {
    marginVertical: 12,
  },
  contactRightPane: {
    flex: 1,
    alignItems: "center",
    alignSelf: "center",
    justifyContent: "center",
    flexDirection: 'row',
    marginHorizontal: 16,
    alignSelf: 'stretch',
    alignItems: 'center',
  },
  contactDesc: {
    flex: 4,
    flexDirection: 'row',
    alignItems: 'center',
    marginLeft: 16
  },
  contactName: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  contactIcon: {
    width: 48,
    height: 48,
    borderRadius: 24,
  },
})
