// hikereactsdk/components/contactPicker/NoSearchResult.js
import React from 'react';
import { StyleSheet, Text, View, ToastAndroid, FlatList, Image, Platform, TouchableOpacity } from 'react-native';
import AppTheme from '../../appthemes/AppTheme';
import ThemeText from '../../appthemes/components/ThemeText';
import CustomTouch from '../CustomTouch';
import { Avatar } from '../Avatar';;

/**
  it renders NoSearchResult.png with text as No results Found in SecondaryColor
*/
export default class NoSearchResult extends React.Component {

  render() {
    return (
      <View style={styles.noSearchResults} >
        <Image style={styles.nosearchresultImage} source={require('../../images/nosearchresults.png')} />
        <ThemeText style={styles.nosearchresultText}
          text={'No results found'}
          color={AppTheme.getColorPalette().secondaryColor}
          fontProfile='FontProfile04'>
        </ThemeText>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  noSearchResults: {
    flex: 1,
    justifyContent: "flex-start",
    marginTop: 24,
    alignItems: 'center'
  },
  nosearchresultImage: {
    height: 121,
    width: 121,
  },
  nosearchresultText: {
    fontSize: 16,
    marginTop: 12,
  }
})
