// hikereactsdk/components/contactPicker/SendButton.js
import React from 'react';
import { StyleSheet, Text, View, ToastAndroid, FlatList, Image, Platform, TouchableOpacity } from 'react-native';
import AppTheme from '../../appthemes/AppTheme';
import ThemeText from '../../appthemes/components/ThemeText';
import CustomTouch from '../CustomTouch';
import { Avatar } from '../Avatar';;
import { ThemeSolidButton } from '../../index';
/**
* sendButton accepts following props
* @property sendButtonEnable
*    boolean for enabling sendButton
* @property onSendPress
*    callback on press of send button
*/
class SendButton extends React.Component {

  constructor(props) {
    super(props);
  }

  render() {
    return (
      <View style={[styles.buttonContainer, {
          backgroundColor: AppTheme.getColorPalette().backgroundColor,
          borderColor: AppTheme.getColorPalette().separatorColor
        }]}>
        <ThemeSolidButton style={{height: 44, margin: 12}}
          onPress={this.props.onSendPress}
          title={this.props.sendButtonText}
          textColorStateProfile='ColorStateProfile05'
          buttonProfile={this.props.sendButtonEnable ? 'ButtonProfile02' : 'ButtonProfile13'}
          fontProfile='FontProfile12'
          />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  buttonContainer: {
    height: 64,
    borderTopWidth: 1,
    alignItems: 'center',
    justifyContent: 'center'
  }
})
export default SendButton;
