// hikereactsdk/components/AppThemeWrapper.js
import React, { Component } from 'react';
import { View, Platform, StyleSheet, Image, Dimensions, StatusBar } from 'react-native';
import AppTheme from '../appthemes/AppTheme';
import LinearGradient from 'react-native-linear-gradient';

const { width, height } = Dimensions.get('window');

/**
 * App Theme Wrapper UI component, use this as a parent to any view where you want to show App Theme background image and rounded corners for chilren
 * @property top
 * 		padding from top
 * @property bottom
 *		padding from bottom
 */
class AppThemeWrapper extends Component {

  constructor(props) {
    super(props);
    this.top=props.top ? props.top :0;
    this.bottom=props.bottom ?props.bottom:0;
  }

  render() {
    var paddingTop = (Platform.OS === 'ios') ? 76 + this.top:
                    ((Platform.Version >= 21) ? (StatusBar.currentHeight + 56) : 56);
    console.log('padding top ' + paddingTop);
    return (
      <View style={[{flex: 1}]}>
        <LinearGradient 
        style={{flex: 1, paddingTop: paddingTop}} 
        colors={AppTheme.getGradientColorsArray()} 
        start={{x: 0.0, y: 0.0}} 
        end={{x: 1.0, y: 1.0}}>
         <Image
         source={{uri: AppTheme.getPatternName()}} 
         style={{tintColor:AppTheme.getPatternAppFillColor(), 
          width: width, height: height, opacity: AppTheme.getPatternAppFillOpacity() ,position:'absolute'}} />
        <View style={[{
          flex: 1,
          borderTopLeftRadius: 4, 
          borderTopRightRadius: 4, 
          marginLeft: 8, 
          marginRight: 8, 
          backgroundColor: AppTheme.getColorPalette().backgroundColor}, 
          this.props.style]}>
          {this.props.children}
        </View>
        </LinearGradient>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  pageCtrMargin:{
    ...Platform.select({
     ios: {
        paddingTop: 76,
     },
     android: {
         paddingTop: 80,
     },
   }),
  }
});

module.exports = {
  AppThemeWrapper
}
