// hikereactsdk/components/Avatar.js
import React, { Component } from 'react';
import { StyleSheet, Text, View, Image, TouchableNativeFeedback, Platform, FlatList } from 'react-native';

let COLOR_PALLETTE = ['#C272F8', '#727BFF', '#01C992', '#3479FF', '#00A4E5', '#00B7C7'];

/**
 * Generic avatar component to show rounded avatar
 * @property size
 * 		container size of avatar
 * @property thumbnail
 *		file path in case of ios and base64 image in android
 * @property iconColor
 *    default icon color if thumbnail is not present
 */
class Avatar extends Component {

  constructor(props) {
    super(props);
    this._getGeneralIcon = this._getGeneralIcon.bind(this);
  }

  _getGeneralIcon(name, iconColor, size) {
    let initial = name ? name.charAt(0) : '#';
    let nameLength = name ? name.length : 0;
    iconColor = iconColor ? iconColor : COLOR_PALLETTE[nameLength % COLOR_PALLETTE.length];

    if(initial === '+')
      initial = '#';
    return (
      <View style={[
        styles.contactIcon,
        {
          backgroundColor: iconColor,
          width: size,
          height: size,
          borderRadius: size/2
        }
        ]}>
        <Text style = {[styles.contactInitials, {fontSize : size / 2}]}>{initial.toUpperCase()}</Text>
      </View>
    )
  }

  render() {
    const profilePicture = this.props.thumbnail ?
    { uri: Platform.OS === "ios"? this.props.thumbnail: `data:image/png;base64, ${decodeURIComponent(this.props.thumbnail)}` }
    : null;

    if (this.props.thumbnail) {
      return (
        <Image source={profilePicture} style={[styles.contactIcon, {
                width: this.props.size,
                height: this.props.size,
                borderRadius: this.props.size/2
              }]}/>
      );
    } else {
      return this._getGeneralIcon(this.props.name, this.props.iconColor , this.props.size);
    }
  }
}

const styles = StyleSheet.create({
  contactIcon: {
    justifyContent: 'center',
    alignItems: 'center'
  },
  contactInitials:{
    color: '#fff',
    opacity: .8,
   // fontFamily: '$fontMedium',
  },
});

module.exports = {
  Avatar
};
