// hikereactsdk/components/FullScreenError.js
import React, { Component } from 'react';
import { StyleSheet, Text, View, Button, ActivityIndicator, Animated, Image } from 'react-native';
import AppTheme from '../appthemes/AppTheme';
import ColorUtils from '../appthemes/ColorUtils';
import ThemeText from '../appthemes/components/ThemeText';
import ThemeOutlineButton from '../appthemes/components/ThemeOutlineButton';

/**
 * Full Screen error UI component which shows error ui and try again button
 * @property tryAgainClicked 
 *        callback method invoked when try again button is clicked
 */
class FullScreenError extends Component {

  /**
   * Renders the full screen error ui component
   */
  render() {
    return (
      <View style={styles.container}>
        <View style={styles.topView}>
          <Image source={require('../images/img_def_empty_nointernet.png')}/>
        </View>
        <View style={styles.bottomView}>
          <ThemeText style={{marginTop: 4, justifyContent: 'center'}}
            color= {AppTheme.getColorPalette().mainColor} 
            fontProfile='FontProfile22'
            text='Something went wrong.'>
          </ThemeText>
          <ThemeText style={{marginTop: 8, justifyContent: 'center'}}
            color= {AppTheme.getColorPalette().secondaryColor} 
            fontProfile='FontProfile05'
            text='Please check your Wi-Fi or mobile data'>
          </ThemeText>
          <ThemeText style={{justifyContent: 'center'}}
            color= {AppTheme.getColorPalette().secondaryColor} 
            fontProfile='FontProfile05'
            text='connection.'>
          </ThemeText>
          <View style={{flexDirection: 'row', marginTop: 32}}>
            <View style={{flex: 0.4}}/>
            <ThemeOutlineButton style={{flex: 1, height: 36}}
              onPress={() => {
                if (this.props.tryAgainClicked) 
                  this.props.tryAgainClicked();
              }} 
              title="TRY AGAIN" 
              textColor={AppTheme.getColorPalette().errorColor}
              buttonProfile='ButtonProfile09'
              fontProfile='FontProfile09'
            />
            <View style={{flex: 0.4}}/>
          </View>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  },
  topView: {
    marginTop: 56,
    height: 128
  },
  bottomView: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'flex-start',
    marginTop: 16,
  }
});

module.exports = {
  FullScreenError
}
