// hikereactsdk/components/HalfScreenError.js
import React, { Component } from 'react';
import { StyleSheet, Text, View, Button, ActivityIndicator, Animated, Image } from 'react-native';
import AppTheme from '../appthemes/AppTheme';
import ColorUtils from '../appthemes/ColorUtils';
import ThemeText from '../appthemes/components/ThemeText';
import ThemeOutlineButton from '../appthemes/components/ThemeOutlineButton';

/**
 * Half Screen error UI component which shows error ui and try again button
 * @property tryAgainClicked 
 *        callback method invoked when try again button is clicked
 */
class HalfScreenError extends Component {

  /**
   * Renders the half screen error ui component
   */
  render() {
    return (
      <View style={styles.container}>
          <Image source={require('../images/img_def_empty_nointernet_3.png')}/>
          <ThemeText style={{marginTop: 4, justifyContent: 'center'}}
            color= {AppTheme.getColorPalette().mainColor} 
            fontProfile='FontProfile04'
            text='Something went wrong.'>
          </ThemeText>
          <ThemeText style={{marginTop: 4, justifyContent: 'center'}}
            color= {AppTheme.getColorPalette().secondaryColor} 
            fontProfile='FontProfile10'
            text='Please check your Wi-Fi or mobile data connection.'>
          </ThemeText>
          <View style={{flexDirection: 'row', marginTop: 24}}>
            <View style={{flex: 0.4}}/>
            <ThemeOutlineButton style={{flex: 1, height: 36}}
              onPress={() => {
                if (this.props.tryAgainClicked) 
                  this.props.tryAgainClicked();
              }} 
              title="TRY AGAIN" 
              textColor={AppTheme.getColorPalette().errorColor}
              buttonProfile='ButtonProfile09'
              fontProfile='FontProfile09'
            />
            <View style={{flex: 0.4}}/>
          </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  },
});

module.exports = {
  HalfScreenError
}
