// hikereactsdk/appthemes/FontStyles.js
import {Platform} from 'react-native';

const fontFaktSoftProNormal = (Platform.OS === 'ios') ? "FaktSoftPro-Normal" : "FaktSoftProNormal"
const fontFaktSoftProMedium = (Platform.OS === 'ios') ? "FaktSoftPro-Medium" : "FaktSoftProMedium"
const fontFaktSoftProBlond = (Platform.OS === 'ios') ? "FaktSoftPro-Blond" : "FaktSoftProBlond"
const fontFaktSoftProSemiBold = (Platform.OS === 'ios') ? "FaktSoftPro-SemiBold" : "FaktSoftProSemiBold"

const fontNeutraDispMedium = (Platform.OS === 'ios') ? "NeutraDispMedium" : "NeutraDispMedium"
const fontNeutraDispTitling = (Platform.OS === 'ios') ? "NeutraDispTitling" : "NeutraDispTitling"

// TODO all caps
/**
 * This file exports the FontStyles object which defines all the design font profiles
 * Each font profile has font size and font family used at runtime for texts
 */
export default FontStyles = {
	FontProfile01 : {
		fontSize: 32,
		fontFamily: fontFaktSoftProMedium,
	},
	FontProfile02 : {
		fontSize: 18,
		fontFamily: fontFaktSoftProMedium,
	},
	FontProfile03 : {
		fontSize: 18,
		fontFamily: fontFaktSoftProNormal,
	},
	FontProfile04 : {
		fontSize: 16,
		fontFamily: fontFaktSoftProMedium,
	},
	FontProfile05 : {
		fontSize: 16,
		fontFamily: fontFaktSoftProNormal,
	},
	FontProfile06 : {
		fontSize: 16,
		fontFamily: fontFaktSoftProBlond,
	},
	FontProfile07 : {
		fontSize: 14,
		fontFamily: fontFaktSoftProMedium,
	},
	FontProfile08 : {
		fontSize: 14,
		fontFamily: fontFaktSoftProBlond,
	},
	FontProfile09 : {
		fontSize: 12,
		fontFamily: fontFaktSoftProMedium,
	},
	FontProfile10 : {
		fontSize: 12,
		fontFamily: fontFaktSoftProNormal,
	},
	FontProfile11 : {
		fontSize: 16,
		fontFamily: fontFaktSoftProMedium,
		// todo all caps
	},
	FontProfile12 : {
		fontSize: 14,
		fontFamily: fontFaktSoftProMedium,
		// todo all caps
	},
	FontProfile13 : {
		fontSize: 12,
		fontFamily: fontFaktSoftProMedium,
		// todo all caps
	},
	FontProfile14 : {
		fontSize: 16,
		fontFamily: fontFaktSoftProNormal,
		// todo all caps
	},
	FontProfile15 : {
		fontSize: 14,
		fontFamily: fontFaktSoftProNormal,
		// todo all caps
	},
	FontProfile16 : {
		fontSize: 12,
		fontFamily: fontFaktSoftProNormal,
		// todo all caps
	},
	FontProfile17 : {
		fontSize: 10,
		fontFamily: fontFaktSoftProMedium,
	},
	FontProfile18 : {
		fontSize: 10,
		fontFamily: fontFaktSoftProNormal,
	},
	FontProfile19 : {
		fontSize: 14,
		fontFamily: fontFaktSoftProNormal,
	},
	FontProfile20 : {
		fontSize: 12,
		fontFamily: fontFaktSoftProBlond,
	},
	FontProfile21 : {
		fontSize: 10,
		fontFamily: fontFaktSoftProMedium,
		// todo all caps
	},
	FontProfile22 : {
		fontSize: 20,
		fontFamily: fontFaktSoftProMedium,
		// todo all caps
	},
	FontProfile23 : {
		fontSize: 56,
		fontFamily: fontNeutraDispTitling,
	},
	FontProfile24 : {
		fontSize: 24,
		fontFamily: fontNeutraDispTitling,
	},
	FontProfile25 : {
		fontSize: 30,
		fontFamily: fontNeutraDispMedium,
	},
	FontProfile26 : {
		fontSize: 24,
		fontFamily: fontNeutraDispMedium,
	},
	FontProfile27 : {
		fontSize: 32,
		fontFamily: fontFaktSoftProBlond,
	},
	FontProfile28 : {
		fontSize: 18,
		fontFamily: fontFaktSoftProBlond,
	},
	FontProfile29 : {
		fontSize: 40,
		fontFamily: fontFaktSoftProBlond,
	},
	FontProfile30 : {
		fontSize: 22,
		fontFamily: fontFaktSoftProMedium,
	},
	FontProfile31 : {
		fontSize: 36,
		fontFamily: fontFaktSoftProNormal
	},
	FontProfile32: {
		fontSize: 44,
		fontFamily: fontFaktSoftProBlond
	},
	FontProfile33: {
		fontSize: 14,
		fontFamily: fontNeutraDispTitling
	},
	FontProfile34: {
		fontSize: 24,
		fontFamily: fontFaktSoftProBlond
	},
	FontProfile35: {
		fontSize: 24,
		fontFamily: fontFaktSoftProNormal
	},
	FontProfile36: {
		fontSize: 24,
		fontFamily: fontFaktSoftProMedium
	},
	FontProfile37: {
		fontSize: 56,
		fontFamily: fontFaktSoftProNormal
	}
}
