// hikereactsdk/appthemes/TextColorStateStyles.js
import AppTheme from './AppTheme'

let styles;

/**
 * This class defines all the text color state profiles based on app theme.
 * Each text color state profile consists of default color, opacity, selected color which will be used at runtime for texts.
 */
class TextColorStateStyles {

	/**
	 * Initializes styles object which consists of text color state profile.
	 */
	static initStyles() {
		var isNightTheme = AppTheme.isNightTheme();
		var colorPalette = AppTheme.getColorPalette();
		styles = {
			ColorStateProfile01: {
				defaultColor: colorPalette.accentColor,
				pressedOpacity: 0.5,
				selectedColor: isNightTheme ? colorPalette.supportive6Color : colorPalette.secondaryColor,
			},
			ColorStateProfile02: {
				defaultColor: colorPalette.accentColor,
				pressedOpacity: 0.5,
			},
			ColorStateProfile03: {
				defaultColor: colorPalette.mainColor,
				pressedOpacity: 0.5
			},
			ColorStateProfile04: {
				defaultColor: isNightTheme ? colorPalette.mainColor : colorPalette.accentColor,
				pressedOpacity: 0.5,
				disableColor: isNightTheme ? colorPalette.mainColor : colorPalette.hintColor,
			},
			ColorStateProfile05: {
				defaultColor: isNightTheme ? colorPalette.mainColor : colorPalette.backgroundColor,
				pressedOpacity: 0.5,
			},
			ColorStateProfile06: {
				defaultColor: isNightTheme ? colorPalette.tertiaryColor : colorPalette.secondaryColor,
				pressedOpacity: 0.5,
				disableColor: isNightTheme ? colorPalette.hintColor : colorPalette.secondaryColor,
			},
			ColorStateProfile07: {
				defaultColor: isNightTheme ? colorPalette.secondaryColor : colorPalette.hintColor,
				selectedColor: isNightTheme ? colorPalette.mainColor : colorPalette.accentColor,
			},
			ColorStateProfile08: {
				defaultColor: isNightTheme ? colorPalette.tertiaryColor : colorPalette.secondaryColor,
				pressedOpacity: 0.5,
				selectedColor: colorPalette.mainColor
			},
			ColorStateProfile09: {
				defaultColor: colorPalette.secondaryColor,
				pressedOpacity: 0.5,
				selectedColor: colorPalette.accentColor,
			},
			ColorStateProfile10: {
				defaultColor: colorPalette.transBgColor,
				pressedOpacity: 0.5,
			}
		}
	}

	/**
	 * Returns styles object which contains text color state profiles.
	 */
	static getStyles() {
		return styles;
	}
}

export default TextColorStateStyles;
