// hikereactsdk/components/Dialog/Radio.js
import  React, {Component } from 'react';
import { View, Text, Image, TouchableOpacity, StyleSheet } from 'react-native';
import ThemeIcon from '../../appthemes/components/ThemeIcon';
/** Radio accepts following props
* @property selected
*    boolean for showing radio button select/unselect
* @property selectionColor
*    color of the radio button when selected
* @property style
*    style applied to radio button
* @property onPress
*    callback on press of radio button
*/
class Radio extends Component {
  constructor(props) {
    super(props);
  }
  render() {
    const {selectionColor, style, selected, onPress} = this.props;
    return (
      <TouchableOpacity style={style} onPress={onPress} >
        {
          selected ?
          <View style={[styles.radioButton, selectionColor]}>
            <ThemeIcon
              onPress={onPress}
              style={styles.iconStyle}
              iconProfile='IconProfile18'
              source={require('../../images/ic_med_tick.png')} />
          </View>
          :
          <View style={[styles.radioButton, styles.unselected]}>
            <ThemeIcon
              onPress={onPress}
              style={styles.iconStyle}
              iconProfile='IconProfile18'
              source={require('../../images/ic_med_tick.png')} />
          </View>
        }
      </TouchableOpacity>
    );
  }
}
const styles = StyleSheet.create({
  iconStyle: {
    width: 12,
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  },
  radioButton: {
    height: 22,
    width: 22,
    margin: 2,
    marginRight: 10,
    borderRadius: 11,
    alignItems: 'center',
    justifyContent: 'center',
  },
  unselected: {
    borderWidth: 1.2,
    borderColor: '#333',
  }
});
export default Radio;
