// hikereactsdk/components/hikeContactMultiChooser/hikeContactMultiChooser.js
import React, { Component } from 'react';
import {
	View,
	Text,
	ListView,
	ActivityIndicator,
	StyleSheet,
	Image,
	TouchableNativeFeedback,
	ToastAndroid,
	Alert
} from 'react-native';
import { HikeSharing } from '../../apis';
import MultipleChoice from './multiselect'

const PropTypes = React.PropTypes;

const categoryNameMap = {
  friends: 'Friends',
  onHike: 'On Hike',
  notOnHike: 'Not On Hike',
};

const placeholderIcon = require('./images/contact.png');

const noop = () => {};

export default class HikeContactMultiChooser extends Component {
  static propTypes = {
		onSelectContact: PropTypes.func,
    showsVerticalScrollIndicator: PropTypes.bool,
    selectedContacts: PropTypes.array,
	};

  static defaultProps = {
    onSelectContact: noop,
    showsVerticalScrollIndicator: false,
	};

  constructor(props) {
    super(props);

    this.ds = new ListView.DataSource({
      rowHasChanged: (r1, r2) => r1 !== r2,
      sectionHeaderHasChanged: (s1, s2) => s1 !== s2,
    });

		this.selectedContacts = this.props.selectedContacts ? JSON.parse(JSON.stringify(this.props.selectedContacts)) : [];

    this.state = {
			selectedContacts: this.props.selectedContacts ? JSON.parse(JSON.stringify(this.props.selectedContacts)) : [],
    };

    this._renderSectionHeader = this._renderSectionHeader.bind(this);
    this._onSelectContact = this._onSelectContact.bind(this);
		this._updateSelectedOptions = this._updateSelectedOptions.bind(this);
		this.getSelectedContacts = this.getSelectedContacts.bind(this);
  }

	_onSelectContact(contact) {
    this.props.onSelectContact(contact);
		contact.isSelected = !contact.isSelected;
		this.setState({ selectedContacts: this.state.selectedContacts.concat([contact]) });
	}

	_resetContacts() {
		this.setState({ selectedContacts: [] });
		this.selectedContacts = [];
	}

	_updateSelectedOptions() {
 		this.selectedContacts = this.refs.mc.state.selectedOptions;
		this.props.onSelectContact();
	}

	getSelectedContacts() {
		return this.selectedContacts;
	}

  _renderSectionHeader(sectionData, category) {
    if(sectionData.length>0)
      return (
        <View>
          <View style={styles.listSeparator}>
            <View style={styles.emptyLeftSection}></View>
            <View style={styles.borderRow}></View>
          </View>
          <View style={styles.sectionHeader}>
            <Text style={styles.contactCategory}>{categoryNameMap[category].toUpperCase()}</Text>
          </View>
       </View>
      );
    return null;
  }

  render() {
    return (
      <View style={[styles.ctr, this.props.style]}>
				<MultipleChoice ref="mc" options={this.props.contactsList} selectedOptions={this.selectedContacts}
					maxSelectedOptions={1} onSelection={(option)=> {this._updateSelectedOptions() }} getGeneralIcon = {(contact) => this.props.getGeneralIcon(contact)}/>
			</View>
    );
  }
}

const iconSize = 48;

const styles = StyleSheet.create({
  loader: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  ctr: {
    flex: 1,
    justifyContent: 'flex-start',
  },
  sectionHeader:{
      justifyContent:'center',
      backgroundColor:'#ffffff'
  },
  listSeparator: {
    flex: 1,
    flexDirection: 'row',
    flexWrap: 'wrap',
  },

  emptyLeftSection: {
    width: 82,
    borderBottomWidth: 1,
    borderBottomColor: '#ffffff',
  },

  borderRow: {
    flex: 1,
    borderBottomWidth: 1,
    borderBottomColor: 'rgba(0,0,0,0.06)',
  },

  contactCategory: {
    fontWeight: '300',
    alignSelf:'center',
    marginVertical: 18,
    fontSize:12,
    color:'rgba(0,0,0,0.3)'
  },
  contactItem: {
    flex:1,
    flexDirection: 'row',
    alignItems: 'flex-start',
    paddingLeft:20,
    paddingTop:12,
    paddingBottom:12,
    backgroundColor:'#ffffff',
    alignItems: 'center',
  },
  contactIcon: {
    width: 48,
    height: 48,
    borderRadius: 48,
  },
  contactDesc:{
    flex:1,
    flexDirection:'column',
    alignItems: 'center',
    justifyContent: 'center',
    marginLeft: 16
  },
  contactName: {
    flex:1,
    flexDirection:'row',
    alignSelf:'flex-start',
    fontSize: 16,
    color:'#3D475B',
    marginBottom:6
  },
  contactNumber:{
    flex:1,
    flexDirection:'row',
    alignSelf:'flex-start',
    fontSize: 14,
    color:'rgba(61,71,91,0.5)'
  }
});
