// hikereactsdk/apis/hikeSharing.js
/**
 * Created by anurag on 20/12/16.
 */

import {NativeModules} from 'react-native';

const HikeSharingModule = NativeModules.HikeSharingModule;

/**
 * Hike Sharing Module
 */
class HikeSharing {
  /**
   * Opens contact chooser of Hike App.
   * @param {Object} [params = {}] Parameters for controlling contact chooser behavior.
   * @param {Object} params.showAllContacts False by default which opens the Hike Contact chooser which
   * lists only friends added by a user. Pass true to list all contacts including people who may not
   * be on hike.
   */
  static openContactChooser(params = {}) {
    const showAllContacts = params.showAllContacts;

    const passParams = Object.assign({}, { showAllContacts: false }, showAllContacts ? {showAllContacts} : null);
    HikeSharingModule.openContactChooser(passParams);
  }

  static openGroupCreator(){
      return HikeSharingModule.openGroupCreator();
  }
  /**
   * Get Hike Contacts data with information for friends, people on hike and people not on hike.
   * @param {object} [params] Optional parameters object.
   * @param {Array} [params.msisdns] Optional list of msisdns to pass if specific data is required for some numbers.
   * @returns {Promise.Array} Returns a Promise which when resolved returns an array of contacts.
   */
  static getContactsData(params = {}) {
    return HikeSharingModule.getContactsData(params);
  }
  /**
   * Get list of all Groups with information for group id, name, display pic.
   * @returns {Promise.Array} Returns a Promise which when resolved returns an array of contacts.
   */
	static getGroupsData(params = {}) {
		return HikeSharingModule.getGroupsData();
	}

  /**
   * Opens Contact picker and forwards the cardObject with the selected contacts.
   * @param {object} cardObject.
   * @param {string} hikeMessage message displayed on the Home Screen Chat List.
   * @param {object} sharedData
   * Sample Use :
   * cardObj {is_group_first:true,h:200,ld:{msisdn:"+hikepoll+"},hd:{pollId:id},parent_msisdn:"+hikepoll+",push:"silent",notifText:"Poll Request"}
   * shared Data {cd:{"randomField":0}}
   * hm = "Poll Request"
   * @returns {Promise.Array} Returns a Promise which when resolved returns an array of contacts.
   */
  static sendSharedMessageFromContactPicker(cardObject,hikeMessage,sharedData,react){
    return HikeSharingModule.sendSharedMessageFromContactPicker(cardObject,hikeMessage,sharedData,react);
  }
  /**
   * Get Hike Contacts data with information for friends, people on hike and people not on hike.
   * @param {string} groupId groupId of the selected group.
   * @returns {Promise.Array} Returns a Promise which when resolved returns an array of contacts.
   */
  static getGroupInfo(groupId){
    return HikeSharingModule.getGroupInfo(groupId);
  }

  /**
   * Forward web card to chat.
   *
   * @param params  the params
   */
  static forwardWebCardToChat(params) {
    return HikeSharingModule.forwardWebCardToChat(params);
  }

  static getAllEventsData() {
    return HikeSharingModule.getAllEventsData();
  }

  static getAllEventsForMessageHash(messageHash) {
    return HikeSharingModule.getAllEventsForMessageHash(messageHash);
  }

  static getSharedEventsData() {
    return HikeSharingModule.getSharedEventsData();
  }

  static sendNormalEvent(messageHash, eventData) {
    return HikeSharingModule.sendNormalEvent(messageHash, eventData);
  }

  static deleteEvent(eventId) {
    return HikeSharingModule.deleteEvent(eventId);
  }

  static deleteAllEventsForMessage(messageHash) {
    return HikeSharingModule.deleteAllEventsForMessage(messageHash);
  }

  static getSelfContactInfo() {
    return HikeSharingModule.getSelfContactInfo();
  }

  /**
    * Call this method to post a status update to timeline.
    *
    * @param status        : Status Text Need to be share with Image
    * @param moodId        : Pass
    *                      "-1" if no mood
    *                      "0", happy
    *                      "1", sad
    *                      "2", in_love
    *                      "3", surprised
    *                      "4", confused
    *                      "5", angry
    *                      "6", sleepy
    *                      "7", hungover
    *                      "8", chilling
    *                      "9", studying
    *                      "10", busy
    *                      "11", love
    *                      "12", middle_finger
    *                      "13", boozing
    *                      "14", movie
    *                      "15", caffeinated
    *                      "16", insomniac
    *                      "17", driving
    *                      "18", traffic
    *                      "19", late
    *                      "20", shopping
    *                      "21", gaming
    *                      "22", coding
    *                      "23", television
    *                      "33", music
    *                      "34", partying_hard
    *                      "35", singing
    *                      "36", eating
    *                      "37", working_out
    *                      "38", cooking
    *                      "39", beauty_saloon
    *                      "40", sick
    * @param imageFilePath : Path of the image on the client. Image should only be of jpeg format and compressed.
    *                      <p>
    *                      Status = null, moodId = -1 & imageFilePath = null should not hold together
   */
  static postStatusUpdate(status, moodId, imageFilePath) {
    return HikeSharingModule.postStatusUpdate(status, moodId, imageFilePath);
  }

  /**
   * React Bridge version 8
   *
   * Call this method to post a status update to timeline with reactView screenshot as image.
   * This method has two variants first we can post screenshot with edit image using isEditImageRequire
   * flag as true or can directly post image completely in background using flag false.
   *
   * @param status             : Status Text Need to be share with Image
   * @param moodId             : Pass
   *                           -1 if no mood
   *                           0, happy
   *                           1, sad
   *                           2, in_love
   *                           3, surprised
   *                           4, confused
   *                           5, angry
   *                           6, sleepy
   *                           7, hungover
   *                           8, chilling
   *                           9, studying
   *                           10, busy
   *                           11, love
   *                           12, middle_finger
   *                           13, boozing
   *                           14, movie
   *                           15, caffeinated
   *                           16, insomniac
   *                           17, driving
   *                           18, traffic
   *                           19, late
   *                           20, shopping
   *                           21, gaming
   *                           22, coding
   *                           23, television
   *                           33, music
   *                           34, partying_hard
   *                           35, singing
   *                           36, eating
   *                           37, working_out
   *                           38, cooking
   *                           39, beauty_saloon
   *                           40, sick
   * @param isEditImageRequire : true if edit image screen require, false if need to post image on timeline completely
   *                           in background
   * @param imageCropData      : Readable map having data to crop image
   *                           param with key x for value, x coordinate of the first pixel in source
   *                           param with key y for value, y coordinate of the first pixel in source
   *                           param with key width for value, number of pixels in each row
   *                           param with key height for value, he number of rows
   */
  static postToTimelineWithCropOption(status, moodId, isEditImageRequire, imageCropData) {
    return HikeSharingModule.postToTimelineWithCropOption(status, moodId, isEditImageRequire, imageCropData);
  }

  static shareTextExternal(title, text) {
    return HikeSharingModule.shareTextExternal(title, text);
  }


  /***
   * Get details of list of usersList
   *
   * @param usersList : Array of users [{uid: '3ddddf33'}]
  */
  static getUserDetailsV2(usersList) {
    return HikeSharingModule.getUserDetailsV2(usersList);
  }

  static sendOrUpdateCard(params) {
    return HikeSharingModule.sendOrUpdateCard(params);
  }

  /**
  * Check if Hidden mode is active or not
  */
  static stealthModeIsActive() {
    return HikeSharingModule.stealthModeIsActive();
  }

  /**
   * Calling this function will initiate forward of the message to a friend or group.
   *
   * @param json        : the card object data for the forwarded card. This data will be the card object for the new forwarded card
   *                    that'll be created. The platform version of the card should be same as the bot, that is defined by the server. The
   *                    app name and app package will also be added from the card object of the bot metadata.
   * @param hikeMessage : the hike message to be included in notif tupple and conversation tupple.
   */
  static forwardMessageToChat(json, hikeMessage) {
    return HikeSharingModule.forwardMessageToChat(json, hikeMessage);
  }

  /**
   *
   * @param json :  cardObj json data, having details with key
   *             timeline (boolean) :
   *             appName (String) : Optional having default value : hikecricketGC
   *             appPackage (String) :  Optional having default value : https://s3-ap-southeast-1.amazonaws.com/platform-qa-segments/micro-apps/hikecricketGC/v1/hikecricketGC.zip
   *             mAppVersionCode (int) :  Optional having default value :  1
   *             target_platform (int) : Optional having default value : same with botInfo.getMetadata()
   * @param hikeMessage : the hike message to be included
   */
  static forwardBotToChat(json, hikeMessage) {
    return HikeSharingModule.forwardBotToChat(json, hikeMessage);
  }

  /**
   * Get friends data
   */
  static getFriendsData() {
    return HikeSharingModule.getFriendsData();
  }
  static postcardToTimeline(ssm){
    return HikeSharingModule.postcardToTimeline(ssm);
  }
  static sendOrUpdateReactCard(params){
    return HikeSharingModule.sendOrUpdateReactCard(params)
  }

  static sendSharedMessage(params){
    return HikeSharingModule.sendSharedMessage(params)
  }
  
}

export default HikeSharing;
